core/src/main/scala/kafka/controller/ReplicaStateMachineV2.scala

```diff
@@ -0,0 +1,318 @@
+package kafka.controller
+
+import java.util.Properties
+
+import kafka.api.LeaderAndIsr
+import kafka.common.{StateChangeFailedException, TopicAndPartition}
+import kafka.log.LogConfig
+import kafka.utils.Logging
+import org.apache.zookeeper.KeeperException
+import org.apache.zookeeper.KeeperException.Code
+
+import scala.collection.mutable
+
+class ReplicaStateMachineV2(controller: KafkaController, zookeeperClient: ZookeeperClient,
+                            controllerChannelManager: ControllerChannelManager) extends Logging {
+  private val controllerContext = controller.controllerContext
+  private val controllerId = controller.config.brokerId
+  private val replicaState: mutable.Map[PartitionAndReplica, ReplicaState] = mutable.Map.empty
+  private val stateChangeLogger = KafkaController.stateChangeLogger
+  private val controllerBrokerRequestBatch = new ControllerBrokerRequestBatch(controller)
+
+  this.logIdent = "[Replica state machine on Controller " + controller.config.brokerId + "]: "
+
+  def handleStateChanges(replicaId: Int, partitions: Seq[TopicAndPartition], targetState: ReplicaState,
+                         callbacks: Map[TopicAndPartition, Callbacks]): Unit = {
+    if (partitions.nonEmpty) {
+      try {
+        controllerBrokerRequestBatch.newBatch()
+        doHandleStateChanges(replicaId, partitions, targetState, callbacks)
+        controllerBrokerRequestBatch.sendRequestsToBrokers(controller.epoch)
+      } catch {
+        case e: Throwable => error("Error while moving some replicas to %s state".format(targetState), e)
+      }
+    }
+  }
+
+  private def doHandleStateChanges(replicaId: Int, partitions: Seq[TopicAndPartition], targetState: ReplicaState,
+                                   callbacks: Map[TopicAndPartition, Callbacks]): Unit = {
+    val replicas = partitions.map(partition => PartitionAndReplica(partition.topic, partition.partition, replicaId))
+    val (validReplicas, invalidReplicas) = replicas.partition(replica => isValidTransition(replica, targetState))
+    invalidReplicas.foreach(replica => logInvalidTransition(replica, targetState))
+    targetState match {
+      case NewReplica =>
+        val getDataResponses = KafkaControllerZkUtils.getTopicPartitionStates(zookeeperClient, validReplicas.map(replica => replica.topicAndPartition -> replica).toMap)
```

Comment:
Yeah I noticed this as well. I mainly just wanted to keep the existing behavior the same and worry about tweaking the existing behavior in later patches.

The only reason I can come up with for this zk lookup is to notice if another controller takes over while this current controller is in the process of doing this transition.

Reply:
Hmm, since we are fixing the issue with ZK session expiration, there shouldn't be more than 1 controller accessing ZK at the same time. In general, I agree that we don't want to make major changes to Partition/ReplicaStateMachine while refactoring ZK accesses. However, if this simplifies how we use ZK, it may be worth doing.